integration/spark2/src/main/scala/org/apache/spark/sql/execution/datasources/SparkCarbonTableFormat.scala

```diff
@@ -509,30 +514,17 @@ private class CarbonOutputWriter(path: String,
 
   // TODO Implement writesupport interface to support writing Row directly to recordwriter
   def writeCarbon(row: InternalRow): Unit = {
-    val data = new Array[AnyRef](fieldTypes.length + partitionData.length)
-    var i = 0
-    val fieldTypesLen = fieldTypes.length
-    while (i < fieldTypesLen) {
-      if (!row.isNullAt(i)) {
-        fieldTypes(i) match {
-          case StringType =>
-            data(i) = row.getString(i)
-          case d: DecimalType =>
-            data(i) = row.getDecimal(i, d.precision, d.scale).toJavaBigDecimal
-          case other =>
-            data(i) = row.get(i, other)
-        }
-      }
-      i += 1
-    }
+    val totalLength = fieldTypes.length + partitionData.length
+    val data: Array[AnyRef] = CommonUtil.getObjectArrayFromInternalRowAndConvertComplexType(row,
+      fieldTypes,
+      totalLength)
     if (partitionData.length > 0) {
-      System.arraycopy(partitionData, 0, data, fieldTypesLen, partitionData.length)
+      System.arraycopy(partitionData, 0, data, fieldTypes.length, partitionData.length)
```

Comment:
Is `fieldTypes.length` equal to number of non-partition columns?

Reply:
yes